pkg/protocol/block_dispatcher.go

```diff
@@ -0,0 +1,345 @@
+		protocol.HookInitialized(b.initNetworkConnection)
+		protocol.HookStopped(b.shutdown)
+	})
+}
+
+// Dispatch dispatches the given block to the correct engine instance.
+func (b *BlockDispatcher) Dispatch(block *model.Block, src network.PeerID) error {
+	slotCommitment := b.protocol.ChainManager.LoadCommitmentOrRequestMissing(block.ProtocolBlock().SlotCommitmentID)
+	if !slotCommitment.SolidEvent().WasTriggered() {
+		if !b.unsolidCommitmentBlocks.Add(slotCommitment.ID(), types.NewTuple(block, src)) {
+			return ierrors.Errorf("failed to add block %s to unsolid commitment buffer", block.ID())
+		}
+
+		return ierrors.Errorf("failed to dispatch block %s: slot commitment %s is not solid", block.ID(), slotCommitment.ID())
+	}
+
+	matchingEngineFound := false
+	for _, engine := range []*engine.Engine{b.protocol.MainEngineInstance(), b.protocol.CandidateEngineInstance()} {
+		if engine != nil && (engine.ChainID() == slotCommitment.Chain().ForkingPoint.ID() || engine.BlockRequester.HasTicker(block.ID())) {
+			if !b.inWarpSyncRange(engine, block) {
+				engine.ProcessBlockFromPeer(block, src)
+			}
+
+			matchingEngineFound = true
+		}
+	}
+
+	if !matchingEngineFound {
+		return ierrors.Errorf("failed to dispatch block %s: no matching engine found", block.ID())
+	}
+
+	return nil
+}
+
+// initEngineMonitoring initializes the automatic monitoring of the engine instances.
+func (b *BlockDispatcher) initEngineMonitoring() {
+	b.monitorLatestEngineCommitment(b.protocol.mainEngine)
+
+	b.protocol.engineManager.OnEngineCreated(b.monitorLatestEngineCommitment)
+
+	b.protocol.Events.ChainManager.CommitmentPublished.Hook(func(chainCommitment *chainmanager.ChainCommitment) {
+		chainCommitment.SolidEvent().OnTrigger(func() {
+			b.runTask(func() {
+				b.injectUnsolidCommitmentBlocks(chainCommitment.Commitment().ID())
+			}, b.dispatchWorkers)
+
+			b.runTask(func() {
+				// warpsync only if the observed commitment is at least two commitments ahead.
+				targetEngine := b.targetEngine(chainCommitment)
+				if targetEngine != nil && chainCommitment.Commitment().Index() > targetEngine.Storage.Settings().LatestCommitment().Index()+1 {
+					b.warpSync(targetEngine, chainCommitment)
+				}
+			}, b.warpSyncWorkers)
+		})
+	})
+
+	b.protocol.Events.Engine.Notarization.LatestCommitmentUpdated.Hook(func(commitment *model.Commitment) {
+		b.runTask(func() {
+			b.injectUnsolidCommitmentBlocks(commitment.ID())
+		}, b.dispatchWorkers)
+	})
+
+	b.protocol.Events.Engine.SlotGadget.SlotFinalized.Hook(b.evict)
+}
+
+// initNetworkConnection initializes the network connection of the BlockDispatcher instance.
+func (b *BlockDispatcher) initNetworkConnection() {
+	b.protocol.Events.Engine.BlockRequester.Tick.Hook(func(blockID iotago.BlockID) {
+		b.runTask(func() {
+			b.protocol.networkProtocol.RequestBlock(blockID)
+		}, b.dispatchWorkers)
+	})
+
+	b.pendingWarpSyncRequests.Events.Tick.Hook(func(id iotago.CommitmentID) {
+		b.runTask(func() {
```

Comment:
Why can't we use the workerpool functionality of the event?

Reply:
Because we only want to hand over these tasks to the workerpool if it wasnt shut down yet, so we need to make the queuing conditional.

In general, I think we should eventually move away from providing the workerpool as an option in the event and always queue manually.

```
myEvent.Hook(func() {
    wp.Submit(func() {
        // dosth ...
    }),
)
```

isn't orders of magnitude more verbose as

```
myEvent.Hook(func() {
   // dosth ...
}, event.WithWorkerPool(wp))
```
 and it would allow us to massively simplify the events moving forward, which also makes them faster as we no longer need to allocate additional closures inside the event.